text/0123-pending-code-as-storage-location-for-runtime-upgrades.md

```diff
@@ -0,0 +1,60 @@
+Furthermore, this RFC proposes to introduce `system_version: 3`. The `system_version` was introduced in [`RFC42`](https://polkadot-fellows.github.io/RFCs/approved/0042-extrinsics-state-version.html). Version `3` would then enable the usage of `:pending_code` when applying a runtime code upgrade. This way, the feature can be introduced first and enabled later when the majority of the nodes have upgraded.
+## Drawbacks
+
+Because the first block built with the new runtime code will move the runtime code from `:pending_code` to `:code`, the runtime code will need to be loaded. This means the runtime code will appear in the proof of validity of a parachain for the first block built with the new runtime code. Generally this  is not a problem as the runtime code is also loaded by the parachain when setting the new runtime code.
+There is still the possibility of having state that is not migrated even when following the proposal as presented by this RFC. The issue is that if the amount of data to be migrated is too big, not all of it can be migrated in one block, because either it takes more time than there is assigned for a block or parachains for example have a fixed budget for their proof of validity. To solve this issue there already exist multi-block migrations that can chunk the migration across multiple blocks. Consensus-critical data needs to be migrated in the first block to ensure that block production etc., can continue. For the other data being migrated by multi-block migrations the migrations could for example expose to the outside which keys are being migrated and should not be indexed until the migration is finished.
+
+## Testing, Security, and Privacy
+
+Testing should be straightforward and most of the existing testing should already be good enough. Extending with some checks that `:pending_code` is moved to `:code`.
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+The performance should not be impacted besides requiring loading the runtime code in the first block being built with the new runtime code.
+
+### Ergonomics
+
+It only alters the way blocks are produced and imported after applying a runtime upgrade. This means that only nodes need to be adapted to the changes of this RFC.
+
+### Compatibility
+
+The change will require that the nodes are upgraded before the runtime starts using this feature. Otherwise they will fail to import the block build by `:pending_code`. 
+For Polkadot/Kusama this means that also the parachain nodes need to be running with a relay chain node version that supports this new feature. Otherwise the parachains will stop producing/finalizing nodes as they can not sync the relay chain any more.
```

Comment:
Should the RFC also describe how the feature will be enabled ? Is it by simply bumping system_version to 3 in the runtime once we know all the nodes have upgraded ?

Reply:
> Is it by simply bumping system_version to 3 in the runtime once we know all the nodes have upgraded ?

Yes exactly. This was my idea.

Reply:
Waiting for relay chain validators to upgrade seems like an acceptable way to introduce a breaking change, but parachain nodes are known to lag behind and stay on an older version. Would it be possible to roll out this change on the relay chain side in a backwards compatible manner for parachain nodes/collators? They might even prefer to disable runtime APIs on runtime upgrades to this change.

Reply:
I don't see any way in doing this in a backwards compatible way. Whatever we do here to fix it, we will always need to touch the node AFAIK.

Reply:
I don't think it is practical or desired to require parachain teams to upgrade their collators. This will delay the deployment of this fix for a long time. 

I think the most practical solution is #3 from https://github.com/paritytech/polkadot-sdk/issues/64 .  My argument here is that currently we are doing it wrong, we are bulding block X (which contains `set_code(B)`) with runtime A, but then we call runtime APIs on it using runtime B. We should always call runtime APIs at a block using the same runtime version we used to build it.

Reply:
> My argument here is that currently we are doing it wrong, we are bulding block X (which contains `set_code(B)`) with runtime A, but then we call runtime APIs on it using runtime B. We should always call runtime APIs at a block using the same runtime version we used to build it.

This is the reason this RFC exist to fix this issue. 

> I think the most practical solution is #3 from [paritytech/polkadot-sdk#64](https://github.com/paritytech/polkadot-sdk/issues/64) .

This would require that all the external tooling is doing the same.

Reply:
> [paritytech/polkadot-sdk#64 (comment)](https://github.com/paritytech/polkadot-sdk/issues/64#issuecomment-1691523854) also see the comments here around pruning.

Pruning should not be a problem IMO, on validators/full nodes we don't actually need to call runtime APIs on blocks that old.

> 
> This would require that all the external tooling is doing the same.

Can you explain a bit how is this affected? (I don't see any details in that comment) 
Doesn't the tooling rely on a node to get the data, so if the node is upgraded then it would work fine ?

Reply:
> Pruning should not be a problem IMO, on validators/full nodes we don't actually need to call runtime APIs on blocks that old

So you know all the access patterns of every single app interfacing with the node? I don't think so

Reply:
I don't but it's very easy to fix - not prune the parent of the blocks containing the code upgrade.